File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.py
def solution(phone_book):

    dic = {}
    for phone_number in phone_book:
        dic[phone_number] = 1
    
    for phone_number in phone_book:
        prefix = ""
        for digit in phone_number:
            prefix += digit
            if prefix in dic and prefix != phone_number:
                return False
    
    return True
    

File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.java
import java.util.*;

class Solution {
    public boolean solution(String[] phone_book) {
        Set<String> set = new HashSet<>();
        
        for (String number : phone_book)  {
            set.add(number);
        }
        
        for (String number : phone_book)  {
            for (int i=1; i<number.length(); i++) {
                if (set.contains(number.substring(0, i))) {
                    return false;
                }
            }
        }
        
        return true;
    }
}